rados/read_op_omap_get_vals_by_keys.go

```diff
@@ -88,30 +91,24 @@ func (s *ReadOpOmapGetValsByKeysStep) Next() (*OmapKeyValue, error) {
 //  PREVIEW
 //
 // Implements:
-//  void rados_read_op_omap_get_vals_by_keys(rados_read_op_t read_op,
-//                                           char const * const * keys,
-//                                           size_t keys_len,
-//                                           rados_omap_iter_t * iter,
-//                                           int * prval)
+//  void rados_read_op_omap_get_vals_by_keys2(rados_read_op_t read_op,
+//                                            char const * const * keys,
+//                                            size_t num_keys,
+//                                            const size_t * key_lens,
+//                                            rados_omap_iter_t * iter,
+//                                            int * prval)
 func (r *ReadOp) GetOmapValuesByKeys(keys []string) *ReadOpOmapGetValsByKeysStep {
 	s := newReadOpOmapGetValsByKeysStep()
 	r.steps = append(r.steps, s)
 
-	cKeys := make([]*C.char, len(keys))
-	defer func() {
-		for _, cKeyPtr := range cKeys {
-			C.free(unsafe.Pointer(cKeyPtr))
-		}
-	}()
-
-	for i, key := range keys {
-		cKeys[i] = C.CString(key)
-	}
+	cKeys := cutil.NewBufferGroupStrings(keys)
+	defer cKeys.Free()
 
-	C.rados_read_op_omap_get_vals_by_keys(
+	C.rados_read_op_omap_get_vals_by_keys2(
 		r.op,
-		&cKeys[0],
+		(**C.char)(cKeys.CStringsPtr()),
```

Comment:
Hmm... so, before, when we used `rados_read_op_omap_get_vals_by_keys()`, we *had* to make a copy of the strings, because they needed to be null-terminated. But now, with `rados_read_op_omap_get_vals_by_keys2()` we don't even have to make a copy. We only need a length array. But yeah, because of pointer-passing-rules we would need to use PtrGuard for the `BufferGroup` like we did here: https://github.com/ceph/go-ceph/blob/72493ed2f744bf80f457978daa2918f120bfc53d/internal/cutil/sync_buffer.go#L18

So, yeah, let's keep it like that, it's internal and not performance critical. Maybe you can add a TODO comment to use PtrGuard later, when we clean the other PtrGuard-stuff up?

Reply:
Let's keep it like it is, because PtrGuard is still optional, and you would have to implement twi versions. We can change later to copy-less version, if it simplifies the code. (Performance is no argument in this case, I think.) Just add a TODO, so we can find it later.

Reply:
@ansiwen alright, will do. I'm just still puzzled by this:

> > because of pointer-passing-rules we would need to use PtrGuard
>
> Even when the C functions make copy of the data and none of the pointers passed to the functions are kept around after they return?

Reply:
Embedded Go pointers in Cgo arguments are not automatically pinned, and therefore forbidden, it would crash. PtrGuard can circumvent that restriction.

Reply:
Aaah, ok I see now, thank you! These `PtrGuard`s would "live" in `BufferGroup` struct right? Is it enough to leave a comment just there? Apologies for rather basic questions...

Reply:
Doesn't matter where.